spring-integration-core/src/main/java/org/springframework/integration/transformer/FromProtobufTransformer.java

```diff
@@ -0,0 +1,157 @@
+import org.springframework.messaging.Message;
+import org.springframework.util.Assert;
+import org.springframework.util.ClassUtils;
+import org.springframework.util.ConcurrentReferenceHashMap;
+
+/**
+ * A Protocol Buffer transformer to instantiate {@link com.google.protobuf.Message} objects from {@code byte[]}.
+ *
+ * @author Christian Tzolov
+ * @since 6.1
+ */
+public class FromProtobufTransformer extends AbstractTransformer implements BeanClassLoaderAware {
+
+	private static final Map<Class<?>, Method> methodCache = new ConcurrentReferenceHashMap<>();
+
+	private final Class<? extends com.google.protobuf.Message> defaultType;
+
+	private ClassLoader beanClassLoader;
+
+	private Expression typeIdExpression = new FunctionExpression<Message<?>>(
+			(message) -> message.getHeaders().get(ProtoHeaders.TYPE));
+
+	private EvaluationContext evaluationContext;
+
+	/**
+	 * Construct an instance with the supplied default type to create.
+	 * @param defaultType the type.
+	 */
+	public FromProtobufTransformer(Class<? extends com.google.protobuf.Message> defaultType) {
+		Assert.notNull(defaultType, "'defaultType' must not be null");
+		this.defaultType = defaultType;
+	}
+
+	@Override
+	public void setBeanClassLoader(ClassLoader classLoader) {
+		this.beanClassLoader = classLoader;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type. Default {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpression(Expression expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = expression;
+	}
+
+	/**
+	 * Set the expression to evaluate against the message to determine the type id. Default
+	 * {@code headers['proto_type']}.
+	 * @param expression the expression.
+	 */
+	public void setTypeExpressionString(String expression) {
+		assertExpressionNotNull(expression);
+		this.typeIdExpression = EXPRESSION_PARSER.parseExpression(expression);
+	}
+
+	private void assertExpressionNotNull(Object expression) {
+		Assert.notNull(expression, "'expression' must not be null");
+	}
+
+	@Override
+	protected void onInit() {
+		this.evaluationContext = IntegrationContextUtils.getEvaluationContext(getBeanFactory());
+	}
+
+	@SuppressWarnings("unchecked")
+	@Override
+	protected Object doTransform(Message<?> message) {
+		Assert.state(message.getPayload() instanceof byte[], "Payload must be a byte[]");
+		Class<? extends com.google.protobuf.Message> type = null;
+		Object value = this.typeIdExpression.getValue(this.evaluationContext, message);
+		if (value instanceof Class) {
+			type = (Class<? extends com.google.protobuf.Message>) value;
+		}
+		else if (value instanceof String) {
+			try {
+				type = (Class<? extends com.google.protobuf.Message>) ClassUtils.forName((String) value,
+						this.beanClassLoader);
+			}
+			catch (ClassNotFoundException | LinkageError e) {
+				throw new IllegalStateException(e);
+			}
+		}
+		if (type == null) {
```

Comment:
Well, I'm not fully OK with this logic, since it is misleading when end-user expression returns `null`.
but I see that you borrowed it from a `SimpleFromAvroTransformer`.
So, then OK - let it be!